### PR TITLE
Fix completion of local functions in static methods

### PR DESCRIPTION
**Pull request by svick**
Contributes to https://github.com/dotnet/roslyn/issues/38074.

### CODE REVIEW

**Comment by svick** [comment]
I was wondering if this code behaves correctly when it comes to `static` local functions and in the process I may have discovered a much more serious bug, see: https://github.com/dotnet/roslyn/issues/38129. Either that, or I don't understand `static` local functions.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs [hunk 1]
```diff
@@ -1465,7 +1465,7 @@ private static bool IsInstance(Symbol symbol)
                 case SymbolKind.Property:
                 case SymbolKind.Method:
                 case SymbolKind.Event:
-                    return !symbol.IsStatic;
+                    return symbol.RequiresInstanceReceiver();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: >RequiresInstanceReceiver [](start = 34, length = 24)

compiler layer changes should come with compiler layer validation as well. Thanks

Comment (svick) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: Done.

Though the new test does highlight that the naming of `LookupStaticMembers` is misleading: even before this change, it returned local variables too. Maybe at least its documentation should be updated?

Comment (svick) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: Done.

Though the new test does highlight that the naming of `LookupStaticMembers` is misleading: even before this change, it returned local variables too. Maybe at least its documentation should be updated?

**Comment by ivanbasov** [comment]
Thank you, @svick ! Would you like to add a similar change to InvocationExpressionSignatureHelpProvider_MethodGroup.cs ?

**Comment by svick** [comment]
@ivanbasov To clarify, are you saying I should go ahead with the approach I outlined at the end of https://github.com/dotnet/roslyn/issues/38074#issuecomment-522800766: adding the property `RequiresInstanceReceiver` to the public interface `ISymbol`, implement it in all types that implement that interface and then using that in InvocationExpressionSignatureHelpProvider_MethodGroup.cs?

I'm making sure, because I think that's quite a significant change and it's not clear to me from your comments that it's actually what you mean.

**Comment by ivanbasov** [comment]
> @ivanbasov To clarify, are you saying I should go ahead with the approach I outlined at the end of [#38074 (comment)](https://github.com/dotnet/roslyn/issues/38074#issuecomment-522800766): adding the property `RequiresInstanceReceiver` to the public interface `ISymbol`, implement it in all types that implement that interface and then using that in InvocationExpressionSignatureHelpProvider_MethodGroup.cs?
> 
> I'm making sure, because I think that's quite a significant change and it's not clear to me from your comments that it's actually what you mean.

@svick , thank you for clarification! I meant to use `RequiresInstanceReceiver` instead of `!IsStatic` in InvocationExpressionSignatureHelpProvider_MethodGroup.cs

I didn't mean to add it to the interface in the current scope.

**Comment by svick** [comment]
@ivanbasov Like I said in https://github.com/dotnet/roslyn/issues/38074#issuecomment-522800766, it's not that simple: InvocationExpressionSignatureHelpProvider_MethodGroup.cs is in a different project, so it doesn't have access to `RequiresInstanceReceiver`, because that only exists on `internal MethodSymbol`, while IESHP_MG.cs is operating on `IMethodSymbol`.

That's why I suggested adding `RequiresInstanceReceiver` to `ISymbol` (which would make it accessible to IESHP_MG.cs), but that could be seen as an overkill and it's certainly not the only reasonable option. Do you have a suggestion on which approach to take here?

**Comment by ivanbasov** [comment]
> @ivanbasov Like I said in [#38074 (comment)](https://github.com/dotnet/roslyn/issues/38074#issuecomment-522800766), it's not that simple: InvocationExpressionSignatureHelpProvider_MethodGroup.cs is in a different project, so it doesn't have access to `RequiresInstanceReceiver`, because that only exists on `internal MethodSymbol`, while IESHP_MG.cs is operating on `IMethodSymbol`.
> 
> That's why I suggested adding `RequiresInstanceReceiver` to `ISymbol` (which would make it accessible to IESHP_MG.cs), but that could be seen as an overkill and it's certainly not the only reasonable option. Do you have a suggestion on which approach to take here?

Thank you for explanations, @svick !
I have created https://github.com/dotnet/roslyn/issues/38181 to track this issue.

**Comment by 333fred** [comment]
@ivanbasov, looks like there's an interactive test failure. Do we have known flakiness there right now?

**Comment by jcouv** [comment]
To get this in without waiting, I think the PR should target `release/dev16.4-preview1` (I think `master` currently requires some approval process). 
Could you retarget the PR and rebase (or cherry-pick) to ensure only your two commits are in the PR after retargeting?

**Comment by ivanbasov** [comment]
> @ivanbasov, looks like there's an interactive test failure. Do we have known flakiness there right now?

I am not aware if there are current flakiness there. @sharwell may know better.

I have restarted the check execution

**Comment by svick** [comment]
@jcouv Done.

**Comment by jcouv** [comment]
Merged. Thanks @svick!

**Pull request merged**